File: leetcode/53.最大子数组和.go
/*
 * @lc app=leetcode.cn id=53 lang=golang
 *
 * [53] 最大子数组和
 */

// @lc code=start
func maxSubArray(nums []int) int {
	pre, ret := 0, nums[0]
	for i := 0; i < len(nums); i++ {
		pre = max(pre+nums[i], nums[i])
		ret = max(ret, pre)
	}
	return ret
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
// @lc code=end

File: leetcode/53.最大子数组和.rs
/*
 * @lc app=leetcode.cn id=53 lang=rust
 *
 * [53] 最大子数组和
 */

// @lc code=start
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut prev = 0;
        nums.iter().fold(nums[0], |ret, &n| {
            prev = n.max(prev + n);
            ret.max(prev)
        })
    }
}
// @lc code=end
